File: typescript/exercises.ts
import { open } from "node:fs/promises"

export function change(amount: bigint): Map<bigint, bigint> {
  if (amount < 0) {
    throw new RangeError("Amount cannot be negative")
  }
  let counts: Map<bigint, bigint> = new Map()
  let remaining = amount
  for (const denomination of [25n, 10n, 5n, 1n]) {
    counts.set(denomination, remaining / denomination)
    remaining %= denomination
  }
  return counts
}

export function firstThenApply<T,U>(
  items: T[], 
  predicate: (item: T) => boolean, 
  mutator: (item: T) => U
): U | undefined {
  for (const item of items){
    if (predicate(item)){
      return mutator(item)
    }
  }
  return undefined

}


export function* powersGenerator(base:bigint): Generator<bigint>{
  let power = 1n
  while (true){
    yield power;
    power *= base;
  }
}

export async function meaningfulLineCount(filename: string) {
  let count = 0
  const file = await open(filename, 'r')
  for await (const line of file.readLines()) {
    
    const trimmed = line.trim()
    if (trimmed && !trimmed.startsWith('#')) {
      count++
    }
  }
  return count
}


interface Sphere{
  kind: "Sphere"
  radius: number
}
interface Box {
  kind: "Box"
  width: number
  length: number
  depth: number
}
export type Shape = Sphere | Box
export function surfaceArea(shape:Shape): number{
  if (shape.kind == 'Box'){
    const [l,w,d] = [shape.length, shape.width, shape.depth]
    return 2* ((l * w) + (l* d) + (w * d)  )
  }
  return 4 * Math.PI * (shape.radius ** 2)
}
export function volume(shape: Shape): number{
  if (shape.kind == 'Box'){
    return  shape.width * shape.length * shape.depth
  }
  return (4 / 3) * Math.PI * (shape.radius**3)
}


export interface BinarySearchTree<T> {
  size(): number;
  insert(value: T): BinarySearchTree<T>;
  contains(value: T): boolean;
  inorder(): Iterable<T>;
  toString(): string;
}

class Node<T> implements BinarySearchTree<T> {
  constructor(
    public left: BinarySearchTree<T>, 
    public value: T, 
    public right: BinarySearchTree<T>
  ) {}

  size(): number {
    return 1 + this.left.size() + this.right.size();
  }

  insert(value: T): BinarySearchTree<T> {
    if (value < this.value) {
      return new Node(this.left.insert(value), this.value, this.right);
    } else if (value > this.value) {
      return new Node(this.left, this.value, this.right.insert(value));
    }
    return this;
  }

  contains(value: T): boolean {
    if (value === this.value) {
      return true;
    } else if (value < this.value) {
      return this.left.contains(value);
    } else {
      return this.right.contains(value);
    }
  }

  *inorder(): Iterable<T> {
    yield* this.left.inorder();
    yield this.value;
    yield* this.right.inorder();
  }

  toString(): string {
    
    const leftSubstring = this.left.toString();
    const rightSubstring = this.right.toString();
    const leftPart = leftSubstring !== '()' ? leftSubstring : ''
    const rightPart = rightSubstring !== '()' ? rightSubstring : ''
    return `(${leftPart}${this.value}${rightPart})`;
  }
}

export class Empty<T> implements BinarySearchTree<T> {
  size(): number {
    return 0;
  }

  insert(value: T): BinarySearchTree<T> {
    return new Node(new Empty(), value, new Empty());
  }

  contains(value: T): boolean {
    return false;
  }

  *inorder(): Iterable<T> {
    
  }

  toString(): string {
    return `()`;
  }
}

File: ocaml/exercises.ml
exception Negative_Amount

let change amount =
  if amount < 0 then
    raise Negative_Amount
  else
    let denominations = [25; 10; 5; 1] in
    let rec aux remaining denominations =
      match denominations with
      | [] -> []
      | d :: ds -> (remaining / d) :: aux (remaining mod d) ds
    in
    aux amount denominations


let first_then_apply array predicate consumer =
  match List.find_opt predicate array with
  None -> None
| Some x -> consumer x

let powers_generator base =
  let rec generate_from power () = 
    Seq.Cons (power, generate_from  (power * base))
  in
  generate_from  1

  let meaningful_line_count filename =
    let meaningful_line line =
      let trimmed = String.trim line in
      not (String.length trimmed = 0 || String.get trimmed 0 = '#')
    in
    let ic = open_in filename in
    let rec count_lines_aux count =
      try
        let line = input_line ic in
        if meaningful_line line then count_lines_aux (count + 1)
        else count_lines_aux count
      with End_of_file -> count
    in
    Fun.protect ~finally:(fun () -> close_in ic) (fun () -> count_lines_aux 0)
  

    type shape = 
  | Sphere of float
  | Box of float * float * float

let pi = 3.141592653589793

let volume = function 
  | Sphere r -> 4.0 /. 3.0 *. pi *. r ** 3.0
  | Box (l, w, h) -> l *. w *. h

let surface_area = function
  | Sphere r -> 4.0 *. pi *. r ** 2.0
  | Box (l, w, h) -> 2.0 *. (l *. w +. w *. h +. h *. l)

let epsilon = 0.0001

let float_expect actual expected =
  abs_float (actual -. expected) < epsilon


  type 'a binary_search_tree = 
  | Empty
  | Node of 'a * 'a binary_search_tree * 'a binary_search_tree

let rec size tree = 
  match tree with
  | Empty -> 0
  | Node (_, left, right) -> 1 + size left + size right

let rec contains value tree = 
  match tree with
  | Empty -> false
  | Node (v, left, right) -> 
    if value = v then true
    else if value < v then contains value left
    else contains value right

let rec inorder tree = 
  match tree with
  | Empty -> []
  | Node (v, left, right) -> inorder left @ [v] @ inorder right

let rec insert value tree = 
  match tree with
  | Empty -> Node (value, Empty, Empty)
  | Node (v, left, right) ->
    if value = v then tree  
    else if value < v then Node (v, insert value left, right)
    else Node (v, left, insert value right)